locator/token_metadata.cc

```diff
@@ -888,49 +888,34 @@ class tokens_iterator_impl {
     using difference_type = std::ptrdiff_t;
     using pointer = token*;
     using reference = token&;
-private:
-    tokens_iterator_impl(std::vector<token>::const_iterator it, size_t pos)
-    : _cur_it(it), _ring_pos(pos) {}
-
 public:
+    tokens_iterator_impl() = default;
     tokens_iterator_impl(const token& start, const token_metadata_impl* token_metadata)
     : _token_metadata(token_metadata) {
         _cur_it = _token_metadata->sorted_tokens().begin() + _token_metadata->first_token_index(start);
+        _remaining = _token_metadata->sorted_tokens().size();
     }
 
     bool operator==(const tokens_iterator_impl& it) const {
-        return _cur_it == it._cur_it;
-    }
-
-    bool operator!=(const tokens_iterator_impl& it) const {
-        return _cur_it != it._cur_it;
+        return _remaining == it._remaining;
```

Comment:
I checked and it seems ok for iterator comparison *not* to consider the possibility of different containers being compared - see for example https://stackoverflow.com/questions/4657513/comparing-iterators-from-different-containers. So I think your new equality-checking code is correct. But I'm still wondering what was wrong in the old one?

Reply:
Yes, comparing iterators from different containers is undefined by convention. (Side note: this is a fairly direct consequence of the fact that comparing pointers is defined only if they point to the same array). 

As I said in the commit message, the old code was okay. I just thought it was confusing, since I was (wrongly) convinced after first reading that `_ring_pos` doesn't do anything. But it's completely subjective. I can revert the change, if you think the new code is less readable.

I think that the cleanest way to implement this would be to use sentinels (that is: make the end iterator a different type, and write `operator==` only for a `(normal iterator, end iterator)` pair), but that won't work with `boost::iterator_range`. Sentinels were introduced (to `std` and to for-range loops) only in C++20.

Reply:
I still don't understand why this is correct. Wouldn't  `tokens_iterator_impl(1, M) == tokens_iterator_impl(2, M)` be true now and not true in the previous version?

Reply:
Comparison between iterators is defined only for iterators belonging to the same range. `tokens_iterator_impl(T1, M)` and `tokens_iterator_impl(T2, M)` are unrelated and their comparison is meaningless/undefined/incorrect. Only comparisons between related iterators (like the pair returned from `ring_range` and ones obtained by incrementing the first of the pair) is meaningful.